File: tmb_syntax/report.R
## Demonstrate how to pass objects back to R using the REPORT macro
library(Matrix)
library(TMB)

## Compile and load example
compile("report.cpp")
dyn.load(dynlib("report"))

## Generate random objects
set.seed(123)
a <- rnorm(10)
b <- array(rnorm(2*3*4),c(2,3,4))
c <- matrix(rnorm(6),2)
d <- spMatrix(10,15,i=c(1L,6L),j=c(2L,4L),x=3:4)
p <- rnorm(1)
data <- list(a=a,b=b,c=c,d=d)
parameters <- list(p=p)

## REPORT()
## Check objects are unchanged when passed back to R
obj <- MakeADFun(data=data, parameters=parameters, DLL="report")
(rep <- obj$report())
stopifnot( identical(rep$a, a) )
stopifnot( identical(rep$b, b) )
stopifnot( identical(rep$c, c) )
stopifnot( identical(rep$d, d) )
stopifnot( identical(rep$p, p) )
stopifnot( identical(rep$voa, list(c, c)) )
stopifnot( identical(rep$aoa, array(list(c, c), c(1,1,2,1))) )

## ADREPORT()
## Check objects are unchanged when passed back to R
sdr <- sdreport(obj, hessian.fixed=diag(1))
adrep <- as.list(sdr, "Estimate", report=TRUE)
stopifnot( all( adrep$a == a ) )
stopifnot( identical(adrep$b, b) ) ## Dimension preserved
stopifnot( identical(adrep$c, c) ) ## Dimension preserved
stopifnot( all( adrep$p == p ) )

File: tmb_syntax/report.cpp
// Demonstrate how to pass objects back to R using the REPORT macro
#include <TMB.hpp>
template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR        (a);
  DATA_ARRAY         (b);
  DATA_MATRIX        (c);
  DATA_SPARSE_MATRIX (d);
  PARAMETER          (p);

  // cout is supported (but not recommended)
  std::cout << a << "\n";
  std::cout << b << "\n";
  std::cout << c << "\n";
  std::cout << d << "\n";
  std::cout << p << "\n";

  // Report objects back to R:
  REPORT(a);
  REPORT(b);
  REPORT(c);
  REPORT(d);
  REPORT(p);

  //// Vector of anything:
  vector<matrix<Type> > voa(2);
  voa[0] = c;
  voa[1] = c;
  REPORT(voa);

  //// Array of anything:
  vector<int> dim(4);
  dim << 1, 1, 2, 1;
  array<matrix<Type> > aoa(voa, dim);
  REPORT(aoa);

  // AD Report objects back to R:
  ADREPORT(a);
  ADREPORT(b);
  ADREPORT(c);
  ADREPORT(p);

  // AD Reporting transformed objects back to R must also work
  // although will lose dimension.
  ADREPORT(exp(a));
  ADREPORT(exp(b));
  ADREPORT(c.array().exp());
  ADREPORT(exp(p));

  return 0;
}
